indexing-service/src/main/java/org/apache/druid/indexing/input/DruidSegmentReader.java

```diff
@@ -0,0 +1,277 @@
+   */
+  private static class IntermediateRowFromCursorIterator implements Iterator<Map<String, Object>>
+  {
+    private final Cursor cursor;
+    private final BaseLongColumnValueSelector timestampColumnSelector;
+    private final Map<String, DimensionSelector> dimSelectors;
+    private final Map<String, BaseObjectColumnValueSelector> metSelectors;
+
+    public IntermediateRowFromCursorIterator(
+        Cursor cursor,
+        List<String> dimensionNames,
+        List<String> metricNames
+    )
+    {
+      this.cursor = cursor;
+
+      timestampColumnSelector =
+          cursor.getColumnSelectorFactory().makeColumnValueSelector(ColumnHolder.TIME_COLUMN_NAME);
+
+      dimSelectors = new HashMap<>();
+      for (String dim : dimensionNames) {
+        final DimensionSelector dimSelector = cursor
+            .getColumnSelectorFactory()
+            .makeDimensionSelector(new DefaultDimensionSpec(dim, dim));
+        // dimSelector is null if the dimension is not present
+        if (dimSelector != null) {
+          dimSelectors.put(dim, dimSelector);
+        }
+      }
+
+      metSelectors = new HashMap<>();
+      for (String metric : metricNames) {
+        final BaseObjectColumnValueSelector metricSelector =
+            cursor.getColumnSelectorFactory().makeColumnValueSelector(metric);
+        metSelectors.put(metric, metricSelector);
+      }
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+      return !cursor.isDone();
+    }
+
+    @Override
+    public Map<String, Object> next()
+    {
+      final Map<String, Object> theEvent = Maps.newLinkedHashMap();
+      final long timestamp = timestampColumnSelector.getLong();
+      theEvent.put(TimestampSpec.DEFAULT_COLUMN, DateTimes.utc(timestamp));
+
+      for (Entry<String, DimensionSelector> dimSelector : dimSelectors.entrySet()) {
+        final String dim = dimSelector.getKey();
+        final DimensionSelector selector = dimSelector.getValue();
+        final IndexedInts vals = selector.getRow();
+
+        int valsSize = vals.size();
+        if (valsSize == 1) {
+          final String dimVal = selector.lookupName(vals.get(0));
+          theEvent.put(dim, dimVal);
+        } else if (valsSize > 1) {
+          List<String> dimVals = new ArrayList<>(valsSize);
+          for (int i = 0; i < valsSize; ++i) {
+            dimVals.add(selector.lookupName(vals.get(i)));
+          }
+          theEvent.put(dim, dimVals);
+        }
+      }
+
+      for (Entry<String, BaseObjectColumnValueSelector> metSelector : metSelectors.entrySet()) {
+        final String metric = metSelector.getKey();
+        final BaseObjectColumnValueSelector selector = metSelector.getValue();
+        Object value = selector.getObject();
+        if (value != null) {
+          theEvent.put(metric, value);
+        }
+      }
+      cursor.advance();
+      return theEvent;
+    }
+
+    @Override
+    public void remove()
```

Comment:
nit: the `Iterator` interface has the default implementation for `remove()` throwing `UnsupportedOperationException`.

Reply:
Removed the redundant override